File: sarvam_dub.py
"""
Sarvam AI Video Dubbing SDK

A clean Python interface for Sarvam's video dubbing API.

Usage:
    from sarvam_dub import SarvamDubbing

    dub = SarvamDubbing()
    result = dub.dub("video.mp4", target_langs=["Hindi"])
    print(result.dubbed_url)
"""

import os
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, Optional

import requests

BASE_URL = "https://dashboard.sarvam.ai"
API_DUBBING = f"{BASE_URL}/api/dubbing"

LANGUAGES = [
    "Assamese", "Bengali", "English", "Gujarati", "Hindi",
    "Kannada", "Malayalam", "Marathi", "Odia", "Punjabi",
    "Tamil", "Telugu",
]

GENRES = [
    "podcast", "monologue", "advertisement",
    "ott_movie_sequence", "edtech", "academic_lecture_formal",
]


@dataclass
class JobStatus:
    """Current status of a dubbing job."""
    job_id: str
    status: str  # in_progress, completed, failed, partial_failure
    progress: float = 0
    current_step: str = ""
    current_step_label: str = ""
    error_message: str = ""
    dubbed_video_url: str = ""
    original_video_url: str = ""
    duration: float = 0
    file_size: int = 0
    raw: dict = field(default_factory=dict)

    @property
    def is_completed(self) -> bool:
        return self.status == "completed" and bool(self.dubbed_video_url)

    @property
    def is_failed(self) -> bool:
        return self.status in ("failed", "partial_failure")

    @property
    def is_running(self) -> bool:
        return self.status == "in_progress"


@dataclass
class DubResult:
    """Result of a completed dubbing job."""
    job_id: str
    dubbed_url: str
    original_url: str = ""
    duration: float = 0
    file_size: int = 0
    output_path: str = ""


class SarvamDubbingError(Exception):
    """Base exception for Sarvam dubbing errors."""
    pass


class JobCreationError(SarvamDubbingError):
    pass


class UploadError(SarvamDubbingError):
    pass


class JobStartError(SarvamDubbingError):
    pass


class JobFailedError(SarvamDubbingError):
    pass


class SarvamDubbing:
    """
    Sarvam AI Video Dubbing client.

    Usage:
        dub = SarvamDubbing()

        # All-in-one
        result = dub.dub("video.mp4", target_langs=["Hindi"])

        # Step-by-step
        job = dub.create_job(src_lang="English", target_langs=["Hindi"])
        dub.upload(job["upload_url"], "video.mp4")
        dub.start(job["job_id"])
        status = dub.wait(job["job_id"])
        dub.download(status.dubbed_video_url, "output.mp4")
    """

    def __init__(self, base_url: str = BASE_URL, poll_interval: int = 5):
        self.base_url = base_url
        self.api_url = f"{base_url}/api/dubbing"
        self.poll_interval = poll_interval
        self.session = requests.Session()

    def create_job(
        self,
        src_lang: str = "English",
        target_langs: list[str] = None,
        job_name: str = "untitled",
        num_speakers: int = 1,
        genre: str = "podcast",
        editor_flow: bool = False,
    ) -> dict:
        """
        Create a new dubbing job.

        Args:
            src_lang: Source language (e.g., "English")
            target_langs: List of target languages (e.g., ["Hindi", "Tamil"])
            job_name: Name for the job
            num_speakers: Number of speakers in the video
            genre: Content genre (podcast, monologue, etc.)
            editor_flow: Enable editor flow

        Returns:
            dict with job_id, upload_url, srt_upload_url, etc.

        Raises:
            JobCreationError: If job creation fails
        """
        if target_langs is None:
            target_langs = ["Hindi"]

        for lang in [src_lang] + target_langs:
            if lang not in LANGUAGES:
                raise ValueError(f"Invalid language: '{lang}'. Valid: {LANGUAGES}")
        if genre not in GENRES:
            raise ValueError(f"Invalid genre: '{genre}'. Valid: {GENRES}")

        payload = {
            "src_lang": src_lang,
            "target_langs": target_langs,
            "job_name": job_name,
            "num_speakers": num_speakers,
            "genre": genre,
            "editor_flow": editor_flow,
        }

        resp = self.session.post(f"{self.api_url}/jobs", json=payload)
        if resp.status_code != 200:
            raise JobCreationError(f"Failed to create job: {resp.status_code} {resp.text}")

        return resp.json()["data"]

    def upload(self, upload_url: str, video_path: str) -> None:
        """
        Upload a video file to the presigned Azure Blob URL.

        Args:
            upload_url: Presigned URL from create_job response
            video_path: Path to the video file

        Raises:
            FileNotFoundError: If video file doesn't exist
            UploadError: If upload fails
        """
        if not os.path.exists(video_path):
            raise FileNotFoundError(f"Video not found: {video_path}")

        with open(video_path, "rb") as f:
            resp = requests.put(
                upload_url,
                data=f,
                headers={"x-ms-blob-type": "BlockBlob"},
            )

        if resp.status_code not in (200, 201):
            raise UploadError(f"Upload failed: {resp.status_code} {resp.text[:200]}")

    def upload_srt(self, srt_upload_url: str, srt_path: str) -> None:
        """
        Upload an SRT subtitle file.

        Args:
            srt_upload_url: Presigned URL from create_job response
            srt_path: Path to the SRT file
        """
        if not os.path.exists(srt_path):
            raise FileNotFoundError(f"SRT not found: {srt_path}")

        with open(srt_path, "rb") as f:
            resp = requests.put(
                srt_upload_url,
                data=f,
                headers={"x-ms-blob-type": "BlockBlob"},
            )

        if resp.status_code not in (200, 201):
            raise UploadError(f"SRT upload failed: {resp.status_code}")

    def start(self, job_id: str) -> None:
        """
        Start processing a job after upload.

        Args:
            job_id: The job ID from create_job

        Raises:
            JobStartError: If job fails to start
        """
        resp = self.session.post(f"{self.api_url}/jobs/{job_id}/start")
        if resp.status_code != 200:
            raise JobStartError(f"Failed to start job: {resp.status_code} {resp.text}")

    def status(self, job_id: str) -> JobStatus:
        """
        Get current status of a job.

        Args:
            job_id: The job ID

        Returns:
            JobStatus object
        """
        resp = self.session.get(f"{self.api_url}/jobs/{job_id}/live-status")
        if resp.status_code != 200:
            return JobStatus(job_id=job_id, status="unknown")

        data = resp.json().get("data", {})
        export = data.get("export", {})

        return JobStatus(
            job_id=job_id,
            status=data.get("status", "unknown"),
            progress=data.get("progress", 0),
            current_step=data.get("current_step", ""),
            current_step_label=data.get("current_step_label", ""),
            error_message=data.get("error_message", ""),
            dubbed_video_url=export.get("dubbed_video_url", ""),
            original_video_url=export.get("original_video_url", ""),
            duration=export.get("duration", 0),
            file_size=export.get("file_size", 0),
            raw=data,
        )

    def wait(
        self,
        job_id: str,
        on_progress: Optional[Callable[[JobStatus], None]] = None,
        timeout: int = 600,
    ) -> JobStatus:
        """
        Wait for a job to complete, polling at regular intervals.

        Args:
            job_id: The job ID
            on_progress: Optional callback called on each status update
            timeout: Maximum wait time in seconds (default: 600)

        Returns:
            Final JobStatus

        Raises:
            JobFailedError: If the job fails
            TimeoutError: If timeout is exceeded
        """
        start_time = time.time()

        while True:
            if time.time() - start_time > timeout:
                raise TimeoutError(f"Job {job_id} timed out after {timeout}s")

            st = self.status(job_id)

            if on_progress:
                on_progress(st)

            if st.is_completed:
                return st

            if st.is_failed:
                raise JobFailedError(f"Job failed: {st.error_message}")

            time.sleep(self.poll_interval)

    def download(self, dubbed_url: str, output_path: str) -> str:
        """
        Download the dubbed video.

        Args:
            dubbed_url: URL from the completed job status
            output_path: Where to save the file

        Returns:
            The output file path
        """
        if "blob.core.windows.net" in dubbed_url:
            url = f"{self.base_url}/api/media-proxy?url={requests.utils.quote(dubbed_url, safe='')}"
        else:
            url = dubbed_url

        resp = requests.get(url, stream=True)
        if resp.status_code != 200:
            resp = requests.get(dubbed_url, stream=True)

        with open(output_path, "wb") as f:
            for chunk in resp.iter_content(chunk_size=8192):
                f.write(chunk)

        return output_path

    def dub(
        self,
        video_path: str,
        target_langs: list[str] = None,
        src_lang: str = "English",
        num_speakers: int = 1,
        genre: str = "podcast",
        output_path: str = None,
        srt_path: str = None,
        on_progress: Optional[Callable[[JobStatus], None]] = None,
        timeout: int = 600,
    ) -> DubResult:
        """
        All-in-one: create, upload, start, wait, and download.

        Args:
            video_path: Path to video file
            target_langs: Target languages (default: ["Hindi"])
            src_lang: Source language (default: "English")
            num_speakers: Number of speakers (default: 1)
            genre: Content genre (default: "podcast")
            output_path: Output file path (auto-generated if None)
            srt_path: Optional SRT subtitle file path
            on_progress: Optional progress callback
            timeout: Max wait time in seconds

        Returns:
            DubResult with job details and output path

        Example:
            dub = SarvamDubbing()
            result = dub.dub("video.mp4", target_langs=["Hindi", "Tamil"])
            print(f"Dubbed video: {result.output_path}")
        """
        if target_langs is None:
            target_langs = ["Hindi"]

        job_name = Path(video_path).stem

        # Create
        job = self.create_job(
            src_lang=src_lang,
            target_langs=target_langs,
            job_name=job_name,
            num_speakers=num_speakers,
            genre=genre,
        )

        # Upload
        self.upload(job["upload_url"], video_path)

        # Upload SRT if provided
        if srt_path and job.get("srt_upload_url"):
            self.upload_srt(job["srt_upload_url"], srt_path)

        # Start
        self.start(job["job_id"])

        # Wait
        final = self.wait(job["job_id"], on_progress=on_progress, timeout=timeout)

        # Download
        if not output_path:
            stem = Path(video_path).stem
            ext = Path(video_path).suffix
            output_path = f"{stem}_{'_'.join(target_langs)}{ext}"

        self.download(final.dubbed_video_url, output_path)

        return DubResult(
            job_id=job["job_id"],
            dubbed_url=final.dubbed_video_url,
            original_url=final.original_video_url,
            duration=final.duration,
            file_size=final.file_size,
            output_path=output_path,
        )

File: sarvam_dub.js
/**
 * Sarvam AI Video Dubbing SDK (Node.js)
 *
 * Usage:
 *   const { SarvamDubbing } = require("./sarvam_dub");
 *
 *   const dub = new SarvamDubbing();
 *   const result = await dub.dub("video.mp4", { targetLangs: ["Hindi"] });
 *   console.log(result.dubbedUrl);
 */

const fs = require("fs");
const path = require("path");
const https = require("https");
const http = require("http");

const BASE_URL = "https://dashboard.sarvam.ai";

const LANGUAGES = [
  "Assamese", "Bengali", "English", "Gujarati", "Hindi",
  "Kannada", "Malayalam", "Marathi", "Odia", "Punjabi",
  "Tamil", "Telugu",
];

const GENRES = [
  "podcast", "monologue", "advertisement",
  "ott_movie_sequence", "edtech", "academic_lecture_formal",
];

class SarvamDubbingError extends Error {
  constructor(message) { super(message); this.name = "SarvamDubbingError"; }
}

function _request(url, options = {}) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const mod = parsed.protocol === "https:" ? https : http;
    const req = mod.request(parsed, {
      method: options.method || "GET",
      headers: options.headers || {},
    }, (res) => {
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve({
        status: res.statusCode,
        body: Buffer.concat(chunks).toString(),
        headers: res.headers,
      }));
    });
    req.on("error", reject);
    if (options.body) req.write(options.body);
    req.end();
  });
}

function _uploadFile(url, filePath) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const mod = parsed.protocol === "https:" ? https : http;
    const fileSize = fs.statSync(filePath).size;
    const stream = fs.createReadStream(filePath);
    const req = mod.request(parsed, {
      method: "PUT",
      headers: { "x-ms-blob-type": "BlockBlob", "Content-Length": fileSize },
    }, (res) => {
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve({ status: res.statusCode }));
    });
    req.on("error", reject);
    stream.pipe(req);
  });
}

function _downloadFile(url, outputPath) {
  return new Promise((resolve, reject) => {
    const doDownload = (u) => {
      const p = new URL(u);
      const m = p.protocol === "https:" ? https : http;
      m.get(p, (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
          return doDownload(res.headers.location);
        }
        const file = fs.createWriteStream(outputPath);
        res.pipe(file);
        file.on("finish", () => { file.close(); resolve(outputPath); });
      }).on("error", reject);
    };
    doDownload(url);
  });
}

class SarvamDubbing {
  /**
   * @param {Object} options
   * @param {string} [options.baseUrl] - API base URL
   * @param {number} [options.pollInterval] - Poll interval in ms (default: 5000)
   */
  constructor({ baseUrl = BASE_URL, pollInterval = 5000 } = {}) {
    this.baseUrl = baseUrl;
    this.apiUrl = `${baseUrl}/api/dubbing`;
    this.pollInterval = pollInterval;
  }

  /**
   * Create a dubbing job.
   * @param {Object} options
   * @param {string} [options.srcLang="English"]
   * @param {string[]} [options.targetLangs=["Hindi"]]
   * @param {string} [options.jobName="untitled"]
   * @param {number} [options.numSpeakers=1]
   * @param {string} [options.genre="podcast"]
   * @returns {Promise<Object>} Job data with job_id, upload_url, etc.
   */
  async createJob({
    srcLang = "English",
    targetLangs = ["Hindi"],
    jobName = "untitled",
    numSpeakers = 1,
    genre = "podcast",
  } = {}) {
    const resp = await _request(`${this.apiUrl}/jobs`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        src_lang: srcLang,
        target_langs: targetLangs,
        job_name: jobName,
        num_speakers: numSpeakers,
        genre,
        editor_flow: false,
      }),
    });

    if (resp.status !== 200) {
      throw new SarvamDubbingError(`Job creation failed: ${resp.status} ${resp.body}`);
    }
    return JSON.parse(resp.body).data;
  }

  /**
   * Upload video to presigned URL.
   * @param {string} uploadUrl
   * @param {string} videoPath
   */
  async upload(uploadUrl, videoPath) {
    if (!fs.existsSync(videoPath)) {
      throw new Error(`File not found: ${videoPath}`);
    }
    const resp = await _uploadFile(uploadUrl, videoPath);
    if (resp.status !== 200 && resp.status !== 201) {
      throw new SarvamDubbingError(`Upload failed: ${resp.status}`);
    }
  }

  /**
   * Start processing a job.
   * @param {string} jobId
   */
  async start(jobId) {
    const resp = await _request(`${this.apiUrl}/jobs/${jobId}/start`, { method: "POST" });
    if (resp.status !== 200) {
      throw new SarvamDubbingError(`Start failed: ${resp.status} ${resp.body}`);
    }
  }

  /**
   * Get current job status.
   * @param {string} jobId
   * @returns {Promise<Object>} Status object
   */
  async status(jobId) {
    const resp = await _request(`${this.apiUrl}/jobs/${jobId}/live-status`);
    if (resp.status !== 200) return { status: "unknown", jobId };
    const data = JSON.parse(resp.body).data || {};
    const exp = data.export || {};
    return {
      jobId,
      status: data.status || "unknown",
      progress: data.progress || 0,
      currentStep: data.current_step_label || "",
      errorMessage: data.error_message || "",
      dubbedVideoUrl: exp.dubbed_video_url || "",
      originalVideoUrl: exp.original_video_url || "",
      duration: exp.duration || 0,
      fileSize: exp.file_size || 0,
      isCompleted: data.status === "completed" && exp.status === "completed",
      isFailed: ["failed", "partial_failure"].includes(data.status),
      raw: data,
    };
  }

  /**
   * Wait for job completion.
   * @param {string} jobId
   * @param {Object} [options]
   * @param {Function} [options.onProgress] - Callback on each poll
   * @param {number} [options.timeout=600000] - Timeout in ms
   * @returns {Promise<Object>} Final status
   */
  async wait(jobId, { onProgress, timeout = 600000 } = {}) {
    const start = Date.now();
    while (true) {
      if (Date.now() - start > timeout) {
        throw new SarvamDubbingError(`Timeout after ${timeout}ms`);
      }
      const st = await this.status(jobId);
      if (onProgress) onProgress(st);
      if (st.isCompleted) return st;
      if (st.isFailed) throw new SarvamDubbingError(`Job failed: ${st.errorMessage}`);
      await new Promise((r) => setTimeout(r, this.pollInterval));
    }
  }

  /**
   * Download dubbed video.
   * @param {string} dubbedUrl
   * @param {string} outputPath
   * @returns {Promise<string>} Output path
   */
  async download(dubbedUrl, outputPath) {
    let url = dubbedUrl;
    if (dubbedUrl.includes("blob.core.windows.net")) {
      url = `${this.baseUrl}/api/media-proxy?url=${encodeURIComponent(dubbedUrl)}`;
    }
    return _downloadFile(url, outputPath);
  }

  /**
   * All-in-one: create, upload, start, wait, download.
   * @param {string} videoPath
   * @param {Object} [options]
   * @param {string[]} [options.targetLangs=["Hindi"]]
   * @param {string} [options.srcLang="English"]
   * @param {number} [options.numSpeakers=1]
   * @param {string} [options.genre="podcast"]
   * @param {string} [options.outputPath]
   * @param {Function} [options.onProgress]
   * @param {number} [options.timeout=600000]
   * @returns {Promise<Object>} Result with jobId, dubbedUrl, outputPath
   */
  async dub(videoPath, {
    targetLangs = ["Hindi"],
    srcLang = "English",
    numSpeakers = 1,
    genre = "podcast",
    outputPath,
    onProgress,
    timeout = 600000,
  } = {}) {
    const jobName = path.parse(videoPath).name;

    const job = await this.createJob({ srcLang, targetLangs, jobName, numSpeakers, genre });
    await this.upload(job.upload_url, videoPath);
    await this.start(job.job_id);
    const final = await this.wait(job.job_id, { onProgress, timeout });

    if (!outputPath) {
      const ext = path.extname(videoPath);
      outputPath = `${path.parse(videoPath).name}_${targetLangs.join("_")}${ext}`;
    }

    await this.download(final.dubbedVideoUrl, outputPath);

    return {
      jobId: job.job_id,
      dubbedUrl: final.dubbedVideoUrl,
      originalUrl: final.originalVideoUrl,
      duration: final.duration,
      fileSize: final.fileSize,
      outputPath,
    };
  }
}

module.exports = { SarvamDubbing, SarvamDubbingError, LANGUAGES, GENRES };
